tools/spancmp/src/main.rs

```diff
@@ -0,0 +1,289 @@
+//! helper to extract span stats from a chrome trace file
+//! spec: <https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview#heading=h.puwqg050lyuy>
+
+use std::{
+    collections::HashMap,
+    fs::File,
+    io::{BufReader, Read},
+    ops::Div,
+};
+
+use clap::Parser;
+use regex::Regex;
+use serde::Deserialize;
+use termcolor::{Color, ColorChoice, ColorSpec, StandardStream, WriteColor};
+
+/// A span from the trace
+#[derive(Deserialize, Debug)]
+struct Span {
+    /// name
+    name: String,
+    /// phase
+    ph: String,
+    /// timestamp
+    ts: f32,
+}
+
+/// Ignore entries in the trace that are not a span
+#[derive(Deserialize, Debug)]
+struct Ignore {}
+
+/// deserialize helper
+#[derive(Deserialize, Debug)]
+#[serde(untagged)]
+enum SpanOrIgnore {
+    /// deserialize as a span
+    Span(Span),
+    /// catchall that didn't match a span
+    Ignore(Ignore),
+}
+
+/// Trace files are unfinished and not properly formed json
+/// this wrapper finishes the json so that its valid
+struct UnfinishedWrapper {
+    reader: BufReader<File>,
+    buf: Box<[u8; 1]>,
+    finish: String,
+}
+
+impl UnfinishedWrapper {
+    fn from(reader: BufReader<File>) -> UnfinishedWrapper {
+        Self {
+            reader,
+            // last: Ok(0),
+            buf: Box::new([0; 1]),
+            finish: "{}]".to_string(),
+        }
+    }
+}
+
+impl Read for UnfinishedWrapper {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, std::io::Error> {
+        let last = self.reader.read(self.buf.as_mut());
+        if matches!(last, Ok(0)) && self.buf[0] == 10 && !self.finish.is_empty() {
+            let (next, remaining) = self.finish.as_bytes().split_at(1);
+            buf[0] = next[0];
+            self.finish = std::str::from_utf8(remaining).unwrap().to_string();
+            return Ok(1);
+        }
+        buf[0] = self.buf[0];
+        last
+    }
+}
+
+#[derive(Parser, Debug)]
+struct Args {
+    #[clap(short, long, default_value_t = 0.0)]
+    /// Filter spans that have an average shorther than the threshold
+    threshold: f32,
+
+    #[clap(short, long)]
+    /// Filter spans by name matching the pattern
+    pattern: Option<Regex>,
+
+    trace: String,
+    second_trace: Option<String>,
+}
+
+const MARGIN_PERCENT: f32 = 0.02;
```

Comment:
I just hacked this in. It should probably be a CLI argument with a default. :) Later we can base it on the standard deviation of the distributions or something statistically sensible.